lib/IDE/CodeCompletion.cpp

```diff
@@ -4138,10 +4138,46 @@ class CompletionOverrideLookup : public swift::VisibleDeclConsumer {
 
   void addAccessControl(const ValueDecl *VD,
                         CodeCompletionResultBuilder &Builder) {
-    assert(CurrDeclContext->getSelfNominalTypeDecl());
-    auto AccessOfContext =
-        CurrDeclContext->getSelfNominalTypeDecl()->getFormalAccess();
-    auto Access = std::min(VD->getFormalAccess(), AccessOfContext);
+    auto CurrentNominal = CurrDeclContext->getSelfNominalTypeDecl();
+    assert(CurrentNominal);
+
+    auto AccessOfContext = CurrentNominal->getFormalAccess();
+    if (AccessOfContext < AccessLevel::Public)
+      return;
+
+    auto Access = VD->getFormalAccess();
+    // Use the greater access between the protocol requirement and the witness.
+    // In case of:
+    //
+    //   public protocol P { func foo() }
+    //   public class B { func foo() {} }
+    //   public class C: B, P {
+    //     <complete>
+    //   }
+    //
+    // 'VD' is 'B.foo()' which is implicitly 'internal'. But as the overriding
+    // declaration, the user needs to write both 'public' and 'override':
+    //
+    //   public class C: B {
+    //     public override func foo() {}
+    //   }
+    if (Access < AccessLevel::Public &&
+        !isa<ProtocolDecl>(VD->getDeclContext())) {
+      for (auto Conformance : CurrentNominal->getAllConformances()) {
+        auto Proto = Conformance->getProtocol();
+        for (auto Member : Proto->getMembers()) {
+          auto Requirement = dyn_cast<ValueDecl>(Member);
+          if (!Requirement || isa<AssociatedTypeDecl>(Requirement))
+            continue;
+
+          auto Witness = Conformance->getWitnessDecl(Requirement);
+          if (Witness == VD)
+            Access = std::max(Access, Requirement->getFormalAccess());
```

Comment:
The access of the requirement is the same as the access of the protocol, so I believe this entire loop and the search for the witness of 'VD' can go away.

Reply:
You mean like this?
```cpp
    if (Access < AccessLevel::Public &&
        !isa<ProtocolDecl>(VD->getDeclContext())) {
      for (auto Conformance : CurrentNominal->getAllConformances()) {
        auto Proto = Conformance->getProtocol();
        Access = std::max(Access, Proto->getFormalAccess());
      }
    }
```
This may cause unnecessary `public` when the type conforms to two or more protocols. Like
```swift
public protocol P1 { }
protocol P2 {
  func bar()
}
public class B {
  func bar() {}
}
public class C: B, P1, P2 {
  <complete>
}
```
In this case when overriding `bar()`, `C.bar()` is not needed to be `public`.